### ISSUE TITLE
Extract WebM Video

### ISSUE THREAD

**Issue by akhilleusuggo** [WWBN/AVideo-Encoder#294]
We need an option like MP4 , where we can opt HD , SD or Low only .
On webm right now it's like we have no choice . The encoder will encode them to the 3 resolutions .
( I'm re-transcoding all my videos to webm , saved around 40% disk space ) , but without this option , I can't disable the MP4 encoder .

I hope you can implement it , also on the advanced site option .

**Comment by DanielnetoDotCom** [comment]
Sorry I do not get it.

the HD , SD or Low should also reflect on the webm. is not working that way?

**Comment by akhilleusuggo** [comment]
No...
You can not chose to what qualities the webm will be encoded .

Also , when you select to encode WebM , does encode also MP4 , even if the MP4 are disabled ...

**Comment by akhilleusuggo** [comment]
@DanielnetoDotCom 
![image](https://user-images.githubusercontent.com/52116577/95355037-c9175700-08cd-11eb-8b2c-1e442e830e94.png)

**Comment by DanielnetoDotCom** [comment]
I just released a new encoder (v3.3) and a new streamer (v9.6) that handle separately webm and MP4 files, also the new encoder create the resolutions dynamically just like we do for HLS

**Comment by DanielnetoDotCom** [comment]
send me any bugs related to it you find.

I have my head up today for this new release.

**Comment by akhilleusuggo** [comment]
So , first you should remove the command line with the `b:v ` . 

![image](https://user-images.githubusercontent.com/52116577/95872615-a70f5000-0d77-11eb-9e88-9e45d27644c9.png)

Basically you're putting all resolutions to be set to 1MB . We should have something like HLS , where we can control how much bitrate for each resolution . 
For MP4 it's fine , since we have the crf ( this feature is strict to crf ) . webm does not really work on the same way as libx264 .

But I'll give it a try and fine something where we can rate limit the bitrate .

**Comment by DanielnetoDotCom** [comment]
so just remove `-b:v 1M` is fine ???

**Comment by akhilleusuggo** [comment]
no , we really need something like HLS ; 

```
$height = self::getResolution($pathFileName);
            $resolutions = array(360, 480, 720, 1080, 1440, 2160);
            $bandwidth = array(600000, 1000000, 2000000, 4000000, 8000000, 12000000);
            //$videoBitrate = array(472, 872, 1372, 2508, 3000, 4000);
```

**Comment by DanielnetoDotCom** [comment]
Ok, 

we have the following formats ...

1. 2160p: 3840x2160
2. 1440p: 2560x1440
3. 1080p: 1920x1080
4. 720p: 1280x720
5. 480p: 854x480
6. 360p: 640x360
7. 240p: 426x240

what would be the best `-b:v` value for each of it?

**Comment by akhilleusuggo** [comment]
> what would be the best -b:v value for each of it?
There's no exact value . For example on my site I upload very often videos 1080p/60fps , is not the same value for 1080/30fps .

I also use Variable framerate option . In case a video is lower then 60 FPS , will not transcode it to 60 FPS . If it's higher then 60 , neither will be transcoded to 60 FPS . I'm using a limiter .


I just need to be able to add a value . The best result quality/bitrate is to bet crf with limiter ( maxrate ) . This way , if the source video does not require higher bitrate , will be low . But if a video want to go for 10mb , will be limited by the maxrate . 

Containing the bitrate, only this way we can achieve the best value disk-space/quality .

**Comment by akhilleusuggo** [comment]
Ok , I can see you have added some variables that I haven't seen before . But look at the HLS ; 

`-c:v h264 -vf scale=-2:{$resolution} -r 24 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -minrate {$minrate}k -crf 23 -maxrate {$maxrate}k -bufsize {$bufsize}k -c:a aac -b:a {$autioBitrate}k -f hls -hls_time 6 -hls_list_size 0 -hls_key_info_file {$destinationFile}keyinfo {$destinationFile}res{$resolution}/index.m3u8`

`-minrate {$minrate}` , `-maxrate {$maxrate}k` . How can I add this values to the MP4/WEBM ?

**Comment by DanielnetoDotCom** [comment]
I have to change the code for that, 

tell me what would be the best variables/code combination, I will add the variables to replace the code.

**Comment by akhilleusuggo** [comment]
If you gonna change the code , make sure you can add an option for the framerate . Videos of 480p or under should have 30FPS maximum . 

On the code , we need something like this that work ; 
https://github.com/WWBN/AVideo-Encoder/blob/b353fac2a1a51d02f0e0f01bc2fc096007bee0fb/objects/Format.php#L425

```
foreach ($resolutions as $key => $value) {
                if ($height >= $value) {
                    $resolution = $value;
                    $rate = $bandwidth[$key] / 1000;
                    $minrate = ($rate * 0.5);
                    $maxrate = ($rate * 1.5);
                    $bufsize = ($rate * 2);
                    $autioBitrate = $audioBitrate[$key];

```

This way , each can adapt it to it's needs. 
I don't know if possible or not , but an array for $frameRate would be very useful . This way , we can encode ( 240 to 20FPS ) , 360-480 to 30 FPS , 720,1080.... to 60 ( if the source got 60 fps )

**Comment by DanielnetoDotCom** [comment]
Yes, an array is possible, you need to add 7 framerates, one for each resolution

1. 2160p: 3840x2160
2. 1440p: 2560x1440
3. 1080p: 1920x1080
4. 720p: 1280x720
5. 480p: 854x480
6. 360p: 640x360
7. 240p: 426x240

But what I most need your help is to see how the FFMPEG code will look like, currently I have this

## HLS 
`-c:v h264 -vf scale=-2:{$resolution} -r 24 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -minrate {$minrate}k -crf 23 -maxrate {$maxrate}k -bufsize {$bufsize}k -c:a aac -b:a {$autioBitrate}k -f hls -hls_time 6 -hls_list_size 0 -hls_key_info_file {$destinationFile}keyinfo {$destinationFile}res{$resolution}/index.m3u8`

## MP4
`-vf scale=-2:{$resolution} -movflags +faststart -preset veryfast -vcodec h264 -acodec aac -strict -2 -b:a {$autioBitrate}k  -max_muxing_queue_size 1024 -y {$destinationFile}`

## WEBM
`-vf scale=-2:{$resolution} -movflags +faststart -preset veryfast -f webm -c:v libvpx -b:v 1M -acodec libvorbis -b:a {$autioBitrate}k  -y {$destinationFile}`

**Comment by akhilleusuggo** [comment]
Something like this : 

### MP4

`-vf scale=-2:{$resolution} -c:v h264 -crf 23 -preset veryfast -r {$frameRate} -minrate {$minrate}k -maxrate {$maxrate}k -bufsize {$bufsize}k -movflags +faststart -c:a aac -strict -2 -b:a {$autioBitrate}k -max_muxing_queue_size 1024 -y {$destinationFile}
`

The webm doesn't differ so much , just different parameters of the codec .

**Comment by itpastor** [comment]
Team,
I really appreciate the tool and the new feature, but is there a way to automatically encode all video as (low) 360p? 
Thanks!

**Comment by akhilleusuggo** [comment]
@itpastor select all this features , except the ones I have highlighted , clear the cache afterward . 

![image](https://user-images.githubusercontent.com/52116577/96321992-098a7980-1020-11eb-8230-cc4c41003973.png)

**Comment by itpastor** [comment]
Got it. However I don’t have the option for ShowOnlyEncoderAutomaticResolutions. I just installed today, but perhaps my version is older?
![EE3F6812-28C3-4B00-B0E0-CBD5AA6296A5](https://user-images.githubusercontent.com/40804331/96323412-fad59f80-0fea-11eb-9387-57300789460b.png)
![1E0683B9-608E-4145-A4B2-6CB54F342E07](https://user-images.githubusercontent.com/40804331/96323450-2d7f9800-0feb-11eb-861f-adaf2d44abc8.png)

**Comment by akhilleusuggo** [comment]
Please use desktop to debug things .

**Comment by akhilleusuggo** [comment]
@DanielnetoDotCom Any update ?
You asked about an example ;

> Yes, an array is possible, you need to add 7 framerates, one for each resolution
> But what I most need your help is to see how the FFMPEG code will look like, currently I have this

This is how it should like ; 



> Something like this :
> 
> ### MP4
> `-vf scale=-2:{$resolution} -c:v h264 -crf 23 -preset veryfast -r {$frameRate} -minrate {$minrate}k -maxrate {$maxrate}k -bufsize {$bufsize}k -movflags +faststart -c:a aac -strict -2 -b:a {$autioBitrate}k -max_muxing_queue_size 1024 -y {$destinationFile} `
> 
> The webm doesn't differ so much , just different parameters of the codec .

**Comment by akhilleusuggo** [comment]
@DanielnetoDotCom I'm still waiting for this . Because this feature will save a lot of disk space and encoding time .
Right now I'm re-transcoding them manually with a bash script ...

**Comment by DanielnetoDotCom** [comment]
Hi

what are the valid/good framerates for each of those?

- 2160p: 3840x2160
- 1440p: 2560x1440
- 1080p: 1920x1080
- 720p: 1280x720
- 480p: 854x480
- 360p: 640x360
- 240p: 426x240

**Comment by akhilleusuggo** [comment]
well , 
240p ; 20FPS
360p ; 30FPS
480p; 30FPS
720,1080,1440,2160 ; Native FPS ( basically or skip the `-r` flag , or give it variable framerate ) .

Would be cool if we could change it , even if from the script .

**Comment by akhilleusuggo** [comment]
@DanielnetoDotCom This feature is really getting costly for high quality videos if they're not limited . Can you please add this ?
You asked about an example , I gave it . When you got some time , please take a look .

**Comment by DanielnetoDotCom** [comment]
If you add {$framerate} in your configuration it will add a framerate rule

For example: -vf scale=-2:{$resolution} -movflags +faststart -preset veryfast -f webm -c:v libvpx -b:v 1M -acodec libvorbis -b:a {$autioBitrate}k {$framerate}  -y {$destinationFile}

**Comment by akhilleusuggo** [comment]
God bless you !
`$videoFramerate = array(20, 30, 30, 0, 0, 0);`  Does the 0 means same as source ?

**Comment by DanielnetoDotCom** [comment]
Yes, if it is 0, it will ignore the `-r` parameter

**Comment by akhilleusuggo** [comment]
@DanielnetoDotCom I think there's you missed the 240p .

With your array structure , the 20FPS will affect the 360p not the 240p , since the 240p is executed alone .

```
// create command
            $value = 240;
            $resolution = $value;
            $minrate = 200;
            $maxrate = 450;
            $bufsize = 600;
            $autioBitrate = 128;

            $command = get_ffmpeg().' -i {$pathFileName} ';
            eval("\$command .= \" $code\";");
```

**Comment by akhilleusuggo** [comment]
Yes , I confirm the array should be changed . to $videoFramerate = array(30, 30, 0, 0, 0, 0);
Then for the static 240p , should be added the framerate 20

**Comment by DanielnetoDotCom** [comment]
you are right, sorry, I will send an update soon.

**Comment by akhilleusuggo** [comment]
@DanielnetoDotCom I have 2 questions , I'll be happy if you would answer them ;

1. If I uncomment this line , will it work ? Can I replace it on the encoder command line like this ? `-vf scale=-2:{$resolution} -b:v {$videoBitrate}` ? Or this line is not used for anything at all ? Because on HLS we have control over bitrate , but not here .

https://github.com/WWBN/AVideo-Encoder/blob/dde38d3f4c6ef5fa8dc08ddc94f0064410f18b17/objects/Format.php#L350

2. For what do we use ''bandwidth'' ? Because I don't get it . We don't have automatic selection , we don't have playlist , and I've tried to use something like this , and doesn't work . `-vf scale=-2:{$resolution} -b:v {$bandwidth}`

https://github.com/WWBN/AVideo-Encoder/blob/dde38d3f4c6ef5fa8dc08ddc94f0064410f18b17/objects/Format.php#L349


And thank you for the implementation . This will save a lot of disk-space and encoding time .

**Comment by DanielnetoDotCom** [comment]
1. no it will not work yet, but if you have suggestions for this we can make it work
2. bandwidth is for HLS automatic bitrate only, it is useless for MP4 or WEBM

**Comment by akhilleusuggo** [comment]
1. I would love to see it . It should be like the HLS option , but differently if we want , much more simple .
For example , we could use dvr encoding ( I mostly use it ) . Should look something like this `ffmpeg -i $input -vf scale=-2:{$resolution} -crf 23 -maxrate {$videoBitrate}*1.5 -bufsize {$videoBitrate}*2 ...... -f mp4 $output `

We have something similar on HLS;

```
foreach ($resolutions as $key => $value) {
                if ($height >= $value) {
                    $resolution = $value;
                    $rate = $bandwidth[$key] / 1000;
                    $minrate = ($rate * 0.5);
                    $maxrate = ($rate * 1.5);
                    $bufsize = ($rate * 2);
                    $autioBitrate = $audioBitrate[$key];
                    $framerate = "";
                    if(!empty($videoFramerate[$key])){
                        $framerate = " -r {$videoFramerate[$key]} ";
                    }
                    eval("\$command .= \" $code\";");
```

**Comment by DanielnetoDotCom** [comment]
Sorry I do not understand.

adaptive bitrates are only for HLS, NOT for MP4 or WEBM

this is not me, I guess it is videojs/HTML5 limitation

**Comment by akhilleusuggo** [comment]
You don't understand , is not adaptive what I'm proposing , is bitrate control .

with the command line I've linked above , you will have control over the bitrate .

`ffmpeg -i $input -vf scale=-2:{$resolution} -crf 23 -maxrate {$videoBitrate}*1.5 -bufsize {$videoBitrate}*2 ...... -f mp4 $output`

The `-maxrate {$videoBitrate}*1.5` will tell ffmpeg what is the maximum bitrate for the resolution ,
`-bufsize {$videoBitrate}*2` will let know ffmpeg what bufsize is required . FFmpeg documentation recommends the bufsize to be twice the average bitrate , that's why in my example I've added `*2` .

**Comment by DanielnetoDotCom** [comment]
I see, what suppose to be the correct array values for $videoBitrate ??

**Comment by akhilleusuggo** [comment]
As you know , there's no correct , just average .

$videoBitrate = array(600, 1200, 2100, 4200, 7500, 12000); 

Remember that the first array is for 360 , not 240 . This values would need to be increased if the source if 60FPS for ( 720,1080,1440,2160) .  

But those values will limit to now use more then bitrate , witch is good . Like a flood protection , even if the uploaders does an upload of a video of 50mb/s , this will limit the encoding to be standard .

The values above are taking examples from twitch/youtube .
